test/qa/lib/dns.py

```diff
@@ -1,45 +1,47 @@
-from lib import logging
-import sh
+import dns.resolver
 
-# returns True when NordVPN DNS servers are used
-def _is_set_to_nord():
-    return "103.86.96.100" and "103.86.99.100" in sh.cat("/etc/resolv.conf")
 
+# Used for test parametrization.
+DNS_NORD = ["103.86.96.100", "103.86.99.100"]
+DNS_NORD_IPV6 = ["2400:bb40:4444::100", "2400:bb40:8888::100"]
 
-# returns True when Threat Protection Lite DNS servers are used
-def _is_set_to_threat_protection_lite():
-    return "103.86.96.96" and "103.86.99.99" in sh.cat("/etc/resolv.conf")
+# Used for test parametrization.
+DNS_TPL = ["103.86.96.96", "103.86.99.99"]
+DNS_TPL_IPV6 = ["2400:bb40:4444::103", "2400:bb40:8888::103"]
 
+# Used for test parametrization, when the same test has to be run for different values of custom dns.
+DNS_CASE_CUSTOM_SINGLE = "2.0.0.0"
+DNS_CASE_CUSTOM_DOUBLE = "2.0.0.1 2.0.0.2"
+DNS_CASE_CUSTOM_TRIPLE = "2.0.0.3 2.0.0.4 2.0.0.5"
+DNS_CASES_CUSTOM = [DNS_CASE_CUSTOM_SINGLE, DNS_CASE_CUSTOM_DOUBLE, DNS_CASE_CUSTOM_TRIPLE]
 
-# returns True when NordVPN DNS servers are used
-def _is_set_to_nord_ipv6():
-    return "2400:bb40:8888::100" and "2400:bb40:4444::100" in sh.cat("/etc/resolv.conf")
+ALL_DNS_CASES = \
+    DNS_NORD + \
+    DNS_NORD_IPV6 + \
+    DNS_TPL + \
+    DNS_TPL_IPV6 + \
+    DNS_CASE_CUSTOM_SINGLE.split(" ") + \
+    DNS_CASE_CUSTOM_DOUBLE.split(" ") + \
+    DNS_CASE_CUSTOM_TRIPLE.split(" ")
 
 
-# returns True when Threat Protection Lite DNS servers are used
-def _is_set_to_threat_protection_lite_ipv6():
-    return "2400:bb40:4444::103" and "2400:bb40:8888::103" in sh.cat("/etc/resolv.conf")
+def is_unset() -> bool:
+    """ returns True when NordVPN DNS servers are used """
+    return all(os_address != address
+        for os_address in dns.resolver.Resolver().nameservers
+            for address in ALL_DNS_CASES)
 
 
-# returns True when NordVPN app has not modified the DNS
-def is_unset():
-    return not (
-        _is_set_to_nord()
-        and _is_set_to_nord_ipv6()
-        and _is_set_to_threat_protection_lite()
-        and _is_set_to_threat_protection_lite_ipv6()
-    )
+def is_set_for(dns_set_in_app: list) -> bool:
+    """ returns True, if NordVPN application has successfuly set and overriden DNS servers in Resolver """
 
+    # DNS Adddresses set in Resolver:
+    dns_set_in_os_addresses = dns.resolver.Resolver().nameservers
 
-def is_set_for(threat_protection_lite, ipv6="off"):
-    logging.log(sh.cat("/etc/resolv.conf"))
-    if threat_protection_lite == "on" and ipv6 == "on":
-        return _is_set_to_threat_protection_lite_ipv6()
+    # Check DNS Addresses that were successfully set in NordVPN app and Resolver, and put them in list:
+    dns_used_from_app = [address for address in dns_set_in_app if address in dns_set_in_os_addresses]
```

Comment:
Wouldn't work to create 2 sets from the lists and compare them?
e.g.: `return set(dns_set_in_os_addresses) == set(dns_set_in_app)`

Reply:
It seems like this could be a beautiful alternative method to check for same traits, e.g. if list lengths and data is the same.

However, I tried a scenario, where NordVPN application sets `103.86.96.96, 103.86.99.99` DNS on the system, and we slip one additional DNS Address `103.86.99.99` to resolver, making `/etc/resolv.conf` look like the following:
```
nameserver 103.86.96.96
nameserver 103.86.99.99
nameserver 103.86.99.99
```

It seems like current implemented solution would return `False` in scenario described above, but solution with two sets returns `True`.

Since current implemented solution covers more cases, I'd like to stay with it :smile:

Reply:
This looks like a bug to me personally. The duplicated addresses I mean.

Reply:
We currently do not have such case, but in case we would have one like this, my current implementation would detect this issue, and assert would fail, that is why I'd prefer to keep my solution

Reply:
yes, I was thinking that the records are unique. :smile:

if you want to check also for duplicates, then you can keep your implementation or maybe `sorted` would work too(e.g. `return sorted(dns_set_in_os_addresses) == sorted(dns_set_in_app)`)

Reply:
Sorry, I misunderstood your example. I thought that this is an actual bug, but it's just a hypothetical situation :D Yes, I agree then that it's better not to use a set, but Marius `sorted` suggestion should work then.

Reply:
Solution using `sorted` seems to be working :smile: Will implement this :smile: